src/middleware/routeHandler.ts

```diff
@@ -0,0 +1,250 @@
+  }
+
+  if (!isGitAvailable) return
+
+  try {
+    await lock(siteName)
+  } catch (err) {
+    next(err)
+    return
+  }
+
+  await routeHandler(req, res, next).catch(async (err: any) => {
+    await unlock(siteName)
+    next(err)
+  })
+
+  try {
+    await unlock(siteName)
+  } catch (err) {
+    next(err)
+  }
+}
+
+export const attachRollbackRouteHandlerWrapper = (routeHandler: any) => async (
+  req: any,
+  res: any,
+  next: any
+) => {
+  const { userSessionData } = res.locals
+  const { siteName } = req.params
+
+  const { accessToken } = userSessionData
+  const { growthbook } = req
+
+  const shouldUseGitFileSystem = !!growthbook?.getFeatureValue(
+    FEATURE_FLAGS.IS_GGS_ENABLED,
+    false
+  )
+
+  const isGitAvailable = await handleGitFileLock(siteName, next)
+  if (!isGitAvailable) return
+  try {
+    await lock(siteName)
+  } catch (err) {
+    next(err)
+    return
+  }
+
+  let originalStagingCommitSha: any
+  let originalStagingLiteCommitSha: any
+
+  if (shouldUseGitFileSystem) {
+    const results = await Promise.all([
+      gitFileSystemService.getLatestCommitOfBranch(siteName, STAGING_BRANCH),
+      gitFileSystemService.getLatestCommitOfBranch(
+        siteName,
+        STAGING_LITE_BRANCH
+      ),
+    ])
+    const [stagingResult, stagingLiteResult] = results
+
+    if (stagingResult.isErr() || stagingLiteResult.isErr()) {
+      await unlock(siteName)
+      if (stagingResult.isErr()) next(stagingResult.error)
+      if (stagingLiteResult.isErr()) next(stagingLiteResult.error)
+      return
+    }
+    originalStagingCommitSha = stagingResult.value.sha
+    originalStagingLiteCommitSha = stagingLiteResult.value.sha
+    if (!originalStagingCommitSha || !originalStagingLiteCommitSha) {
+      await unlock(siteName)
+      if (stagingResult.isErr()) next(stagingResult.error)
+      return
+    }
+    // Unused for git file system, but to maintain existing structure
+    res.locals.githubSessionData = new GithubSessionData({
+      currentCommitSha: "",
+      treeSha: "",
+    })
+  } else {
+    try {
+      const {
+        currentCommitSha: currentStgCommitSha,
+        treeSha: stgTreeSha,
+      } = await getCommitAndTreeSha(siteName, accessToken, STAGING_BRANCH)
+
+      const {
+        currentCommitSha: currentStgLiteCommitSha,
+      } = await getCommitAndTreeSha(siteName, accessToken, STAGING_LITE_BRANCH)
+
+      const githubSessionData = new GithubSessionData({
+        currentCommitSha: currentStgCommitSha,
+        treeSha: stgTreeSha,
+      })
+      res.locals.githubSessionData = githubSessionData
+
+      originalStagingCommitSha = currentStgCommitSha
+      originalStagingLiteCommitSha = currentStgLiteCommitSha
+    } catch (err) {
+      await unlock(siteName)
+      next(err)
+      return
+    }
+  }
+
+  await routeHandler(req, res, next).catch(async (err: any) => {
+    try {
+      if (shouldUseGitFileSystem) {
+        await backOff(() =>
+          convertNeverThrowToPromise(
```

Comment:
why must convert to promise?

Reply:
because the library that we are using (backoff) expects a promise 
this is impt becuase the existence of the throw -> allows for the retry.
never throw never throws -> no retry occurs